### DISCUSSION TITLE
Malt and Cloth Sim Weirdness

### DISCUSSION THREAD

**Discussion by dibli-goost** [bnpr/Malt#513] [Q&A] [answered]
@pragma37 do you know if any part of Malt's code might be affecting Blender's cloth sim cache in weird ways?

I believe the cache is supposed to automatically re-simulate on changes to base geometry, changing the pose on an armature for example. Every once and a while though, Blender will stop re-simulating in every engine *except* Malt (which isn't ideal because of the latency). I can't seem to find a proper way to reproduce the issue, it just randomly happens, disabling BlenderMalt and restarting blender fixes the issue.

This has been happening for quite a while, but I hadn't done any extensive testing to try and find the root cause. It definitely seems to be linked to Malt. Not sure if I should make an issue report, since I can't reproduce it, and don't have any logs or anything to provide.

**Comment by pragma37** [comment]
There's nothing directly related to cloth sim in Malt's code.
From the point of view of the renderer, it's just another mesh.

And, in theory, Malt shouldn't be doing anything if it's not the active engine.

Maybe some of the callbacks could be interfering in some way, or maybe it's a bug on the Blender side and the way Malt retrieves the scene data triggers an update in a way that the built-in engines don't. But it's hard to say.

Just to recap:
- Cloth sim stops working unless Malt is the selected engine (and rendering?).
- It doesn't get fixed if you restart Blender, even if Malt is not the active engine.
- It only gets fixed if you disable the Malt addon and restart Blender.
- Then you can enable Malt again and the cloth sim keeps working everywhere.

Is that right?

If you can share a non-working file and I can reproduce it I can take a look.
Otherwise, it's going to be pretty hard to figure out anything.

**Reply by dibli-goost** [reply]
Yep that's exactly how I would describe it. It will occur for a little bit when I reload the project, but forcibly clearing the cloth sim cache by changing the start frame value on the cloth sim itself will fix it from that point onwards, it will re-simulate as normal.

I think I've found a way to reproduce it on my machine.

Here's a file where it currently happens for me. When I load this file, the only way to resimulate is to forcibly clear the cache. Once clearing the cache, everything works fine until I re-enable BlenderMalt. At which point it caches it and won't re-simulate, from that point, I either have to disable the addon and restart blender, or change the active rendering engine to Malt.

I'm not sure why simply enabling the addon causes the issue here, because in other blend files of mine, it really does seem to just start happening at random. I can consistently cause the bug with the following steps on my machine though.

[cloth sim bug.zip](https://github.com/bnpr/Malt/files/12198677/cloth.sim.bug.zip)

**Reply by pragma37** [reply]
It's working as expected here. 😕


https://github.com/bnpr/Malt/assets/36610999/5a82340a-2de9-4c7f-8ccc-f274b7868f8f

**Reply by dibli-goost** [reply]
I think that's cuz you keyframed it, which forces it to update (It does so on my end as well here). I'm not keyframing any bones, just moving the bone around in the viewport to test stuff on the fly.

Here's a gif of what's occurring on my end.
https://cdn.discordapp.com/attachments/884570930892910595/1135233004038529024/sim_bug.gif

Admittedly, I have been using Malt for so long now, that maybe I've forgotten how it works by default. Maybe it's not supposed to update, and Malt updating at all is the bug, you can see at the tail end of the clip, the sim jumps a bit like it does in Eevee, since I'm not moving any bones, but the moment I move bones with Malt active, it resimulates.

**Reply by pragma37** [reply, marked as answer]
I get that same behavior as yours (when moving the bone in the viewport) even with Malt fully disabled.
I don't know if that's the intended behavior TBH, but if you think it's a bug you could report it in the Blender bug tracker.

In any case, the "bug" on the Malt side seems to be, indeed, forcing a cache update and not the other way around.
